### DISCUSSION TITLE
Unable to login to piccolo_admin in production

### DISCUSSION THREAD

**Discussion by arunk** [piccolo-orm/piccolo_admin#424] [General]
Hi, I managed to setup and get everything running perfectly in my development environment. But when I moved the code to production, I can see the site running, there are no errors in the log files. I have also made changes in production such as adding the allowed_hosts kwarg to the create_admin method, and specified the domain that the site is on. I have also passed `production: True` to `create_admin`, I have also tried to turn `debug: True` just to see if will give some additional information somewhere. But whatever I've tried, I've been unable to login to the site. I'm able to check the users from running `piccolo user ls` and other commands, I've even changed password just to make sure it wasn't a fat finger issue. The user I'm trying to logging as has admin, superadmin and is active. 
The only error I see is a little orange bar at the bottom of the screen that says `Problem logging in` .
Also attaching a screenshot of the error:
<img width="1268" alt="ss-piccolo-admin-login" src="https://github.com/user-attachments/assets/03cde5fc-3554-4d91-9db0-c16bd19c285f">

**Comment by arunk** [comment]
Ok found something in the browser console:
```log
Request failed index.eebd4d8d.js:79:53332
{
  "data": "Client host can't be found.",
  "status": 400,
  "statusText": "Bad Request",
  "headers": {
    "connection": "keep-alive",
    "content-length": "27",
    "date": "Sat, 05 Oct 2024 08:24:05 GMT",
    "server": "nginx/1.18.0 (Ubuntu)"
  },
  "config": {
    "transitional": {
      "silentJSONParsing": true,
      "forcedJSONParsing": true,
      "clarifyTimeoutError": false
    },
    "adapter": [
      "xhr",
      "http",
      "fetch"
    ],
    "transformRequest": [
      null
    ],
    "transformResponse": [
      null
    ],
    "timeout": 0,
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
    "maxContentLength": -1,
    "maxBodyLength": -1,
    "env": {},
    "headers": {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json",
      "X-CSRFToken": "xxx-xxx-xxx-xxx"
    },
    "method": "post",
    "url": "./public/login/",
    "data": "{\"username\":\"arun\",\"password\":\"xxxxx\"}"
  },
  "request": {}
}
Logging in
```
I have crossed out my password, and the tokens, but the information there is all correct. And I'm logging in from a Firefox browser, in case that makes any difference.

**Reply by arunk** [reply]
Also tried Chromium and got the same error.

**Comment by arunk** [comment]
Didn't notice one more console log, it was higher up in the console. AxiosError:
```
{
    "message": "Request failed with status code 401",
    "name": "AxiosError",
    "stack": "AxiosError: Request failed with status code 401\n    at yl (https://z69.in/admin/assets/index.eebd4d8d.js:3:1055)\n    at XMLHttpRequest.u (https://z69.in/admin/assets/index.eebd4d8d.js:3:6125)\n    at dr.request (https://z69.in/admin/assets/index.eebd4d8d.js:5:1965)\n    at async te.fetchUser (https://z69.in/admin/assets/index.eebd4d8d.js:103:30723)",
    "config": {
        "transitional": {
            "silentJSONParsing": true,
            "forcedJSONParsing": true,
            "clarifyTimeoutError": false
        },
        "adapter": [
            "xhr",
            "http",
            "fetch"
        ],
        "transformRequest": [
            null
        ],
        "transformResponse": [
            null
        ],
        "timeout": 0,
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
        "maxContentLength": -1,
        "maxBodyLength": -1,
        "env": {},
        "headers": {
            "Accept": "application/json, text/plain, */*"
        },
        "method": "get",
        "url": "./api/user/"
    },
    "code": "ERR_BAD_REQUEST",
    "status": 401
}
```

**Reply by dantownsend** [reply]
The `Client host can't be found` problem is coming from the rate limiting middleware:

https://github.com/piccolo-orm/piccolo_api/blob/bc7eefa9d63b018cf45853798acbfd0be0978754/piccolo_api/rate_limiting/middleware.py#L165

For some reason the host value can't be read.

This is probably something to do with the Nginx config. Take a look at this:

https://serverfault.com/questions/955533/nginx-as-reverse-proxy-doesnt-pass-client-host

**Reply by dantownsend** [reply]
The nginx config I typically use is really simple, something like:

```
server {
    listen 80 default_server;

    client_max_body_size 5M;

    location / {
        proxy_pass http://localhost:8000;
    }
}
```

**Reply by arunk** [reply]
My site.conf looks like this:
```
 location / {
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-Host $http_host;
        proxy_redirect off;

        if (!-f $request_filename) {
            proxy_pass http://app_server;
            break;
        }
    }
```
Whatever articles I found online said that the variable used to be $host, but it is now $http_host. I also turned error_log to debug in nginx, and I can see the headers in the log file. They are correct:
```
2024/10/05 11:15:52 [debug] 143129#143129: *1 http proxy header:
"POST /admin/public/login/ HTTP/1.0
X-Forwarded-For: 171.76.82.3
X-Real-IP: 171.76.82.3
Host: z69.in
X-Forwarded-Host: z69.in
Connection: close
Content-Length: 39
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br, zstd
Content-Type: application/json
X-CSRFToken: 89c71357-9df9-48d5-8ef5-388cc05a37f0
Origin: https://z69.in
Referer: https://z69.in/admin/
Cookie: csrftoken=89c71357-9df9-48d5-8ef5-388cc05a37f0
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
Priority: u=0

"
```
Could you please tell me what Headers the middleware is looking for? Maybe it is not either Host, or X-Forwarded-Host?

**Reply by arunk** [reply]
It looks like uvicorn is getting the remote address at the socket level - https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/utils.py#L13 rather than from HTTP headers. If this is the case then this shouldn't work in production for anyone, because no one is going to have their ASGI server running on 80 and 443, there is bound to be a reverse proxy sending this information to the ASGI server and workers. I can't be  hundred percent sure that is where the request.client check is coming from, but this is the other bit of code I see - https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/http/h11_impl.py#L96 and that code is calling the functions from utils.py above.

**Reply by dantownsend** [reply]
I'm away from my computer now, but will look later today. The problem is definitely header related somehow.

**Reply by arunk** [reply]
I'm trying to debug the issue, I'll keep posting whatever I find. So I turned on x-forwarded-proto and x-forwarded-scheme on nginx, because this code here seems to want it: https://github.com/encode/uvicorn/blob/6ffaaf7c2f78274889da1572832dd307a8b117d3/uvicorn/middleware/proxy_headers.py#L38 . But that doesn't help. Tried restarting everything, it still didn't work. 

These are the nginx headers now:
Hmm I'm going to try and put little print statements in these uvicorn lilbrary files and see what values its getting for all this. 
```
2024/10/05 11:43:04 [debug] 144489#144489: *43 http proxy header:
"POST /admin/public/login/ HTTP/1.1
X-Forwarded-Scheme: https
X-Forwarded-Proto: https
X-Forwarded-For: 171.76.82.3
X-Real-IP: 171.76.82.3
Host: z69.in
X-Forwarded-Host: z69.in
Connection: close
Content-Length: 39
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br, zstd
Content-Type: application/json
X-CSRFToken: 89c71357-9df9-48d5-8ef5-388cc05a37f0
Origin: https://z69.in
Referer: https://z69.in/admin/
Cookie: csrftoken=89c71357-9df9-48d5-8ef5-388cc05a37f0
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
Priority: u=0

"
```

**Reply by arunk** [reply]
Tried various permutations of this for a bit - https://unix.stackexchange.com/questions/26715/how-can-i-communicate-with-a-unix-domain-socket-via-the-shell-on-debian-squeeze to see if I could listen on the socket instead of gunicorn and see what data and headers were being passed from nginx. But none of it worked.

**Reply by arunk** [reply]
It sounds like this might be the issue - https://github.com/encode/uvicorn/issues/1068 . The issue has been marked as closed, I'm going to try and install uvicorn from github rather than pypi, the issue was just fixed last week, its probably not made it to pypi release yet.

**Reply by arunk** [reply]
Oh nm, that issue has been pushed to pypi, so the fix is apparently in. Except its not, I guess I should reopen the issue.

**Reply by arunk** [reply]
So the follow up is I tried to setup ProxyHeaderMiddleware inside my blacksheep app, but I don't think I'm doing it right. It isn't working. I have asked for help in the Blacksheep matrix channel. Lets see if they can help. Also from this - https://www.uvicorn.org/deployment/#unix-domain-sockets-uds it sounds like there is an issue with Unix Domain Sockets especially when it comes to uvicorn. So maybe I'm best off running it on a TCP/IP port.

**Reply by dantownsend** [reply]
Yes, I would try running it on a port. That has always worked well for me.

**Reply by arunk** [reply]
So just an update on this, running it on a localhost port is working fine. It would be nice to debug this issue further and figure out why it is not working. My investigations lead me to Python's standard module - asyncio and its Transport class. Now its one thing to fork some libraries but obviously it won't be possible to do that with Python itself. It will be a longer process to submit changes. But again I don't know if it is a problem with the asyncio module and the way it handles sockets, or the way that uvicorn is looking for information from the remote connection using the library. Anyway, I think I'll just be happy with running it on a IP address and port and worry about running it on a unix socket some other time. Thanks for your help and time.

**Reply by dantownsend** [reply]
OK, glad you got to the bottom of it. I haven't tried running it using sockets before.

**Reply by arunk** [reply]
From what I understand running servers like this on a domain socket, versus a network socket there is a big difference in performance. The domain socket is able to handle a lot more data in and out. At some point I'll revisit this issue and see if I can figure out what is going on.
